python/CLOUDTRAIL_S3_DATAEVENTS_ENABLED/CLOUDTRAIL_S3_DATAEVENTS_ENABLED.py

```diff
@@ -0,0 +1,488 @@
+        return evaluations
+
+    #No rule parameter specified, default check for all S3 buckets
+    if not valid_rule_parameters:
+        for trail in trail_list:
+            trail_region = trail['HomeRegion']
+            if trail_region == region:
+                if trail['HasCustomEventSelectors']:
+                    event_selector_output = ct_client.get_event_selectors(TrailName=trail['Name'])['EventSelectors']
+                    data_resource_values = get_data_resource_s3_values(event_selector_output)
+                    if isinstance(data_resource_values, bool):
+                        evaluations.append(build_evaluation(event['accountId'], 'COMPLIANT', event))
+                        return evaluations
+
+    #Rule parameter S3BucketName is specified
+    else:
+        for trail in trail_list:
+            trail_region = trail['HomeRegion']
+            if trail_region == region:
+                if trail['HasCustomEventSelectors']:
+                    event_selector_output = ct_client.get_event_selectors(TrailName=trail['Name'])['EventSelectors']
+                    data_resource_values = get_data_resource_s3_values(event_selector_output)
+                    #if logging is enabled for 'arn:aws:s3', return COMPLIANT
+                    if isinstance(data_resource_values, bool):
+                        evaluations.append(build_evaluation(event['accountId'], 'COMPLIANT', event))
+                        return evaluations
+
+                #get bucket names from data resource values to store them in a list
+                for value in data_resource_values:
+                    # value is of the form arn:aws:s3:::test/, [13:-1] strips 'arn:aws:s3:::' and the trailing slash
+                    bucket_name = value[13:-1]
+                    compliant_buckets.append(bucket_name)
+
+        #match rule parameter values v/s buckets with data events logging enabled
+        non_compliant_buckets = (list(set(valid_rule_parameters) - set(compliant_buckets)))
+        #If non_compliant_buckets is empty, logging is enabled for all required buckets
+        if not non_compliant_buckets:
+            evaluations.append(build_evaluation(event['accountId'], 'COMPLIANT', event))
+            return evaluations
+
+        evaluations.append(build_evaluation(event['accountId'], 'NON_COMPLIANT', event, annotation='AWS CloudTrail trails do not log S3 data events for buckets: ' + str(non_compliant_buckets) + '.'))
+        return evaluations
+
+    evaluations.append(build_evaluation(event['accountId'], 'NON_COMPLIANT', event, annotation='No AWS CloudTrail Trail is configured to log data events for Amazon S3.'))
+    return evaluations
+
+def get_all_trails(ct_client):
+    all_trails = []
+
+    trails_list = ct_client.describe_trails()
+    while True:
+        all_trails += trails_list['trailList']
+        if 'NextToken' in trails_list:
+            trails_list = ct_client.describe_trails(NextToken=trails_list['NextToken'])

```

Comment:
Are you sure it works? I don't see the "NextToken" documented in boto: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudtrail.html#CloudTrail.Client.describe_trails

Reply:
We re-used the code from here: https://github.com/awslabs/aws-config-rules/blob/master/python/CLOUDTRAIL_ENABLED_V2/CLOUDTRAIL_ENABLED_V2.py#L236

Since the rule was in the repo, we didn't use additional checks to test this. Checking using CLI, I don't think this would work as this API doesn't support --max-items as well.
```
aws cloudtrail describe-trails --max-items 1
usage: aws [options] <command> <subcommand> [<subcommand> ...] [parameters]
To see help text, you can run:

  aws help
  aws <command> help
  aws <command> <subcommand> help

Unknown options: --max-items, 1
```
We can remove this.